src/BrowserProtocol.js

```diff
@@ -0,0 +1,94 @@
+import off from 'dom-helpers/events/off';
+import on from 'dom-helpers/events/on';
+import invariant from 'invariant';
+
+import createPath from './utils/createPath';
+
+export default class BrowserProtocol {
```

Comment:
is there a specific reason why these (I am using they are the browser|memory|dom|history equivalent) are classes instead of a middleware themselves? who controls the life of the instance?